### ISSUE TITLE
Use io.ascii to parse FITS ASCII tables?

### ISSUE THREAD

**Issue by embray** [astropy/astropy#3430]
As mentioned in other issues, the eventual goal for `io.fits` is to use Astropy `Table` (or a specialized subclass thereof) as the native format in which to return table data read from FITS files (both at the high level `Table.read()` interface, and the lower-level `HDU.data` interface).

Since `io.ascii` already returns its results as `Table` objects, it seems worth investigating using it to parse FITS ASCII tables, rather than reinventing the wheel.  My knowledge of `io.ascii` is still superficial so I don't know if it supports all the features that would be needed to parse FITS ASCII tables.  And if not, I don't know how much effort it would be to add those features over just using a specialized parsers just for FITS.  This is a reminder to look into it.

**Comment by taldcroft** [comment]
The `fixed_width` reader has the machinery you could use for parsing  actual data fields in FITS ASCII table.  Overall I am guessing this is the easy part of reading a FITS ASCII table, but it might not hurt to consolidate to one bit of code.

**Comment by embray** [comment]
There are some funny FORTRAN-based formatting/parsing rules that I'm not sure how to incorporate, but I agree that it seems like the `fixed_width` reader would be a start at least.

**Comment by embray** [comment]
Ooh, it even already handles [inter-column cruft](http://docs.astropy.org/en/v0.4.3/io/ascii/fixed_width_gallery.html#fixedwidthtwoline)  That was going to be my next question.

**Comment by dhomeier** [comment]
Possibly naïve question: is `io.fits` not using `cfitsio` under the hood for ASCII as well as binary files? That would seem to be far more mature than adapting `io.ascii` to FITS' specific Fortran-style flavour.
As an aside, perhaps HEASARC should be contacted to list astropy as Python interface on their [FITSIO page](http://heasarc.gsfc.nasa.gov/docs/software/fitsio/fitsio.html) rather than some package that has last been updated when pyfits was still alive.

**Comment by embray** [comment]
@dhomeier It does not use CFITSIO, no, for reasons.

That page is pretty woefully out of date.  It doesn't even mention PyFITS despite it being older and more mature than the FITS library it does link to (not to mention maintained).

**Comment by taldcroft** [comment]
@embray - my recollection is that PyFITS used to be pure Python at one point, and for this reason I was pretty sure astropy.io.fits doesn't use cfitsio.  So I was surprised to find cextern/cfitsio in astropy.  What's up with that?

**Comment by embray** [comment]
@taldcroft It uses CFITSIO to support the tile compression format, and always has for as long as PyFITS has supported that format.  But only for that, and it's optional.  I would prefer to write support for that format from "scratch" (excepting rewriting the actual compression libs themselves), but it hasn't been worth the extra effort.  (Though it would be necessary to fully fix #2150--compression of non-contiguous data.  At least, to fix that issue without requiring copying of the data.)

**Comment by dhomeier** [comment]
@embray - yes, that was my point about that page. If you want to handle it internally, making use of the existing io.ascii would certainly be reasonable, and I think FITS tables are probably structured enough that the C parser could handle them most of the time.

**Comment by embray** [comment]
Yes, I looked into this earlier, and found that io.ascii already has 90% of what I would need for that =D

**Comment by Cadair** [comment]
I think it could still be a useful avenue for cleanups inside the io.fits code. I don't think anyone is likely to get to it anytime ever, but I don't think that means it should be closed.

**Comment by dhomeier** [comment]
I also still see this as potentially useful. Would be great of course to know what specific approach @embray had in mind as a starting point.
Also we should consider switching right away to the accelerated `np.loadtxt` instead – which will require numpy >= 1.23, but it does not seem likely that would delay this task in any way. 😁

**Comment by saimn** [comment]
I'm not sure this is the best place to put our effort, and if there is much to gain here. ASCII tables are not used a lot AFAIK, and it's not a lot of code and using `io.ascii` (or `np.loadtxt`) would require handling some FITS specific cases. 
But if someone is interested, here is the code involved:
https://github.com/astropy/astropy/blob/d3b5af50f6c1cfa14da0799c8835426d6db1a39e/astropy/io/fits/hdu/table.py#L820-L834
https://github.com/astropy/astropy/blob/d3b5af50f6c1cfa14da0799c8835426d6db1a39e/astropy/io/fits/fitsrec.py#L847-L885
